### PR TITLE
feat: implement patch state operator

### PR DESCRIPTION
**Pull request by markwhitfeld**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #545
Now that PR #635 has been implemented we can start adding the standard state operators.
This `patch` operator is the first of these.

## What is the new behavior?
Introduced the `patch` operator which will allow for deeper patching as well as the ability to easily use state operators for properties within a state object.

I have added a use of this operator to the Integration example application (albeit a bit contrived to have such a deep state).  _Note that I haven't added anything to the docs in this PR so that the docs can be updated on the release of the major version to avoid confusing the users._

I have added a sub package called `@ngxs/store/operators` to encapsulate these operatros into their own area. The alternative to this subpackage would be to add them to a separate package called `@ngxs/state-operators` but this would now give the users an additional install step to take to use these operators, which is not ideal (so we are trying out the sub-package approach for now).

**NOTE:** This PR is to allow us to explore how this sub-package feels to use in the dev branch before committing to an approach.

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

### CODE REVIEW

**Comment by markwhitfeld** [comment]
Tried to fix the code climate issues, but they are caused by differences between what prettier and code climate think are good practice. I have gone with the prettier style enforced during commit.

**Comment by splincode** [comment]
I would also like to have a description in the documentation, for what it may come in handy.

```ts
@Action(SetPrefix)
  public setPrefix({ getState, setState, patchState }: StateContext<TodoStateModel>) {
    setState(
      patch({
        pizza: patch({
          model: patch({
            toppings: (topping: any) => 'Mr. ' + topping
          })
        })
      })
    );
  }
```

What is better than this?

```ts
setState(({ pizza }) => ({ 
    pizza: { 
       model: {  toppings: 'Mr. ' + pizza.model.toppings } 
    })
);
```

I think it would be better to do so merge operator

```ts
// state = { a: { a: 1, b: 2 } }
// b = { a: { c: 3} }
setState(state => merge(state, b)) // { a: { a: 1, b: 2, c: 2 } }
```

```ts
setState(state => replace(state, 'pizza.model.toppings', (toppings) => 'Mr' + toppings))
```

File: integration/app/store/todos/todos.state.ts [hunk 1]
```diff
@@ -22,9 +24,15 @@ export class TodosState {
...
-    patchState({ pizza: { model: { toppings: 'Mr. ' + pizza1 } } });
+    setState(
+      patch({
+        pizza: patch({
+          model: patch({
+            toppings: (topping: any) => 'Mr. ' + topping
+          })
+        })
+      })
+    );
```
Comment (deebloo) [inline integration/app/store/todos/todos.state.ts]: how well does the type inference work when combining patches like this?

Comment (splincode) [reply integration/app/store/todos/todos.state.ts]: I think bad, but maybe we can usage generic for patch<T>

Comment (markwhitfeld) [reply integration/app/store/todos/todos.state.ts]: @deebloo It works very well. It took me 3 hours to get the operator done with tests and about 12 frustrating hours to get the type checking working nicely. Have a look at the bigger examples in the tests. 
You can supply a type T here for the patch operator. Unfortunately this integration example lacked decent types and is deeply nested, not the ideal use case for the patch operator.

I agree that operators like those suggested by @splincode would be better suited for this particular scenario.

**Comment by splincode** [comment]
@markwhitfeld Unfortunately, I can not find an application and I have never seen people do this.

```ts
const original: Model = {
             a: 1,
             b: {
               hello: {
                 person: {
                   name: 'you'
                 }
               },
               goodbye: {
                 person: {
                   name: 'Mark'
                 }
               }
             }
           }
```

I think you just need to use the operator merge

```ts
function mergeDeep(...objects) {
  const isObject = obj => obj && typeof obj === 'object';
  
  return objects.reduce((prev, obj) => {
    Object.keys(obj).forEach(key => {
      const pVal = prev[key];
      const oVal = obj[key];
      
      if (Array.isArray(pVal) && Array.isArray(oVal)) {
        prev[key] = pVal.concat(...oVal);
      }
      else if (isObject(pVal) && isObject(oVal)) {
        prev[key] = mergeDeep(pVal, oVal);
      }
      else {
        prev[key] = oVal;
      }
    });
    
    return prev;
  }, {});
}
```

image

**Pull request merged**